File: Problems/621/main.go
package main

import "fmt"

func main() {
	testcases := []struct {
		tasks []byte
		n     int
	}{
		{
			[]byte{'A', 'C', 'A', 'B', 'D', 'B'},
			1,
		},
	}

	for _, tc := range testcases {
		fmt.Println(leastInterval(tc.tasks, tc.n))
	}
}

func leastInterval(tasks []byte, n int) int {
	freq, maxFreq, mostFreqCnt := make([]int, 26), -1, 0
	for _, t := range tasks {
		freq[t-'A']++
	}
	for _, f := range freq {
		if f > maxFreq {
			maxFreq = f
			mostFreqCnt = 1
		} else if f == maxFreq {
			mostFreqCnt++
		}
	}
	if mostFreqCnt >= n+1 {
		return len(tasks)
	}
	max := func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}
	return len(tasks) + max(0, (maxFreq-1)*(n-mostFreqCnt+1)-len(tasks)+mostFreqCnt*maxFreq)
}

File: Problems/621/main.cc
#include <vector>
#include <iostream>
#include <string>
#include <map>
#include <stack>
#include <algorithm>
#include <numeric>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <functional>
#include <sstream>
#include <set>
#include <cctype>

using namespace std;

class Solution
{
public:
    int leastInterval(vector<char> &tasks, int n)
    {
        n++;
        unordered_map<char, int> cnt;
        for (auto &t : tasks)
            cnt[t]++;

        priority_queue<int, vector<int>> pq;
        for (auto &[_, v] : cnt)
            pq.push(v);

        int res = 0;
        while (!pq.empty())
        {
            int k = min(int(pq.size()), n);

            vector<int> buf;
            res += k;
            for (int i = 0; i < k; i++)
            {
                int freq = pq.top();
                pq.pop();
                if (freq > 1)
                    buf.push_back(freq - 1);
            }
            for (auto &f : buf)
                pq.push(f);
            if (!pq.empty())
                res += n - k;
        }

        return res;
    }
};
